litestar/_openapi/plugin.py

```diff
@@ -0,0 +1,102 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any
+
+from yaml import dump as dump_yaml
+
+from litestar._openapi.datastructures import OpenAPIContext
+from litestar._openapi.path_item import create_path_item_for_route
+from litestar.di import Provide
+from litestar.exceptions import ImproperlyConfiguredException
+from litestar.openapi.config import OpenAPIConfig
+from litestar.openapi.controller import OpenAPINotFoundException
+from litestar.openapi.spec import OpenAPI
+from litestar.plugins import InitPluginProtocol, ReceiveRouteProtocol
+from litestar.response import Response
+from litestar.routes import HTTPRoute
+from litestar.serialization import decode_json, encode_json, get_serializer
+from litestar.status_codes import HTTP_404_NOT_FOUND
+
+if TYPE_CHECKING:
+    from litestar.app import Litestar
+    from litestar.config.app import AppConfig
+    from litestar.routes import BaseRoute
+
+
+class OpenAPIPlugin(InitPluginProtocol, ReceiveRouteProtocol):
+    def __init__(self, app: Litestar) -> None:
```

Comment:
Is there a different way to get the `app` instance into the plugin? If we ever want to expose this to the users, they won't be able to instantiate the plugin without the app. Maybe we can consider adding the `Litestar` instance to the `AppConfig`?

Reply:
I've just refactored the dependency providers to access the type encoders from `request.app` instead of the `self.app` reference, so the only thing it needs the app for at the moment is to get the openapi schema plugins from the registry.

However, the thing standing most in the way of us making the plugin public interface is that order is important at the moment. B/c any of the init plugins, or on_app_init hooks could configure an `OpenAPIConfig` object, we require that this plugin runs after all of the other init plugins/hooks.

However, If `AppConfig.plugins` was an instance of `PluginRegistry` instead of a list, and we removed `openapi_config` from `Litestar` and `AppConfig` so that using the OpenAPI plugin was more like using the other types of plugins, where the plugin receives the config object directly, then I don't think order would matter. Irrespective of when the OpenAPI plugin ran, it could grab a reference to the plugin registry from the app config object and know that it would have access to all of the openapi schema plugins when it needed them.

I wonder if we could do this in a mostly non-breaking way by making the plugin registry subclass from list?

Reply:
I'm happy to leave it as is - at least for now - and if we find a reason to want to make it public eventually then we can revisit.

Reply:
> Also, making the registry mutable would be a breaking change IMO.

I didn't think the registry was a documented thing - its not documented on the `Litestar` class, nor in the `plugins` package (at least for API docs).

Reply:
> I didn't think the registry was a documented thing 

You're right! Well.. It should be! 
Nevertheless, it's exposed via a non-private attribute on the `Litestar` object. I think we should consider that part of our public API.